app/src/main/java/com/nutomic/syncthingandroid/util/FileUtils.java

```diff
@@ -0,0 +1,148 @@
+        String volumePath = getVolumePath(volumeId, context);
+        if (volumePath == null) {
+            return File.separator;
+        }
+        if (volumePath.endsWith(File.separator)) {
+            volumePath = volumePath.substring(0, volumePath.length() - 1);
+        }
+        String documentPath = getDocumentPathFromTreeUri(treeUri);
+        if (documentPath.endsWith(File.separator)) {
+            documentPath = documentPath.substring(0, documentPath.length() - 1);
+        }
+        if (documentPath.length() > 0) {
+            if (documentPath.startsWith(File.separator)) {
+                return volumePath + documentPath;
+            } else {
+                return volumePath + File.separator + documentPath;
+            }
+        } else {
+            return volumePath;
+        }
+    }
+
+    @SuppressLint("ObsoleteSdkInt")
+    @TargetApi(21)
+    private static String getVolumePath(final String volumeId, Context context) {
+        if (!isCompatible) {
+            Log.e(TAG, "getVolumePath called on unsupported API level");
+            return null;
+        }
+        try {
+            StorageManager mStorageManager =
+                    (StorageManager) context.getSystemService(Context.STORAGE_SERVICE);
+            Class<?> storageVolumeClazz = Class.forName("android.os.storage.StorageVolume");
+            Method getVolumeList = mStorageManager.getClass().getMethod("getVolumeList");
+            Method getUuid = storageVolumeClazz.getMethod("getUuid");
+            Method getPath = storageVolumeClazz.getMethod("getPath");
+            Method isPrimary = storageVolumeClazz.getMethod("isPrimary");
+            Object result = getVolumeList.invoke(mStorageManager);
+
+            final int length = Array.getLength(result);
+            for (int i = 0; i < length; i++) {
+                Object storageVolumeElement = Array.get(result, i);
+                String uuid = (String) getUuid.invoke(storageVolumeElement);
+                Boolean primary = (Boolean) isPrimary.invoke(storageVolumeElement);
+                Boolean isPrimaryVolume = (primary && PRIMARY_VOLUME_NAME.equals(volumeId));
+                Boolean isExternalVolume = ((uuid != null) && uuid.equals(volumeId));
+                Boolean isHomeVolume = (uuid == null && HOME_VOLUME_NAME.equals(volumeId));
+                if (isPrimaryVolume || isExternalVolume) {
+                    // Return path if the correct volume corresponding to volumeId was found.
+                    return (String) getPath.invoke(storageVolumeElement);
+                } else if (isHomeVolume) {
+                    // Reading the environment var avoids hard coding the case of the "documents" folder.
+                    return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS).getAbsolutePath();
+                }
+                Log.d(TAG, "Skipping volume, uuid = '" + uuid + "', volumeId = '" + volumeId + "'");

```

Comment:
We should leave this in place. It helped me catch crashes if the user chose from a document provider category like the "home directory" which was not supported - now it's implemented correctly but more foreign apps could hook into this SAF UI place.